### ISSUE TITLE
Allow inpainting pipelines to accept latents only inputs

### ISSUE THREAD

**Issue by patrickvonplaten** [huggingface/diffusers#4815]
`diffusers` allows to keep text-to-image outputs in latent format to be passed directly to image-2-image pipelines which can save a lot of time & memory, especially when pipelines are chained together (such as is the case for sdxl-base + refiner).

Now we should allow the same functionality for inpainting. We have refactored the mask & image processing, but we still can't pass image latents (and mask latents). 

In short the following four code samples should all work both:
- `"runwayml/stable-diffusion-inpainting"`
- and `"runwayml/stable-diffusion-v1-5"`

## 1. Pass only image as latents as mask as non-latents for inpainting checkpoints

```py
from diffusers import AutoPipelineForInpainting
from diffusers.utils import load_image
import torch
torch_device = "cuda"

pipe = AutoPipelineForInpainting.from_pretrained("runwayml/stable-diffusion-inpainting", torch_dtype=torch.float16, variant="fp16", use_safetensors=True) 
# pipe = AutoPipelineForInpainting.from_pretrained("runwayml/stable-diffusion-v1-5", torch_dtype=torch.float16, variant="fp16", use_safetensors=True) 

pipe = pipe.to(torch_device)

img_url = "https://raw.githubusercontent.com/CompVis/latent-diffusion/main/data/inpainting_examples/overture-creations-5sI6fQgYIuo.png"
mask_url = "https://raw.githubusercontent.com/CompVis/latent-diffusion/main/data/inpainting_examples/overture-creations-5sI6fQgYIuo_mask.png"

image = load_image(img_url)
mask_image = load_image(mask_url)

prompt = "dslr photography of an empty bench, high quality"
generator = torch.Generator(device="cuda").manual_seed(0)

image= pipe.image_processor.preprocess(image).to(torch.float16).to(pipe.device)
image = pipe.vae.encode(image).latent_dist.sample() * pipe.vae.config.scaling_factor

image = pipe(
    prompt=prompt,
    image=image,
    mask_image=mask,
    mask_latents=mask_latents, # this one is new
    guidance_scale=8.0,
    num_inference_steps=20,
    generator=generator,
).images[0]
```

## 2. Pass both image as latents and mask as latents

```py
from diffusers import AutoPipelineForInpainting
from diffusers.utils import load_image
import torch
torch_device = "cuda"

pipe = AutoPipelineForInpainting.from_pretrained("runwayml/stable-diffusion-inpainting", torch_dtype=torch.float16, variant="fp16", use_safetensors=True)
# pipe = AutoPipelineForInpainting.from_pretrained("runwayml/stable-diffusion-v1-5", torch_dtype=torch.float16, variant="fp16", use_safetensors=True

pipe = pipe.to(torch_device)

img_url = "https://raw.githubusercontent.com/CompVis/latent-diffusion/main/data/inpainting_examples/overture-creations-5sI6fQgYIuo.png"
mask_url = "https://raw.githubusercontent.com/CompVis/latent-diffusion/main/data/inpainting_examples/overture-creations-5sI6fQgYIuo_mask.png"

image = load_image(img_url)
mask_image = load_image(mask_url)

prompt = "dslr photography of an empty bench, high quality"
generator = torch.Generator(device="cuda").manual_seed(0)

height, width = 512, 512
image= pipe.image_processor.preprocess(image).to(torch.float16).to(pipe.device)
mask = pipe.mask_processor.preprocess(mask_image).to(torch.float16).to(pipe.device)

masked_image = image * (mask < 0.5)

image = pipe.vae.encode(image).latent_dist.sample() * pipe.vae.config.scaling_factor
mask_latents = pipe.vae.encode(masked_image).latent_dist.sample() * pipe.vae.config.scaling_factor
mask = torch.nn.functional.interpolate(mask, size=(height // pipe.vae_scale_factor, width // pipe.vae_scale_factor))

image = pipe(
    prompt=prompt,
    image=image,
    mask_image=mask,
    mask_latents=mask_latents, # this one is new
    guidance_scale=8.0,
    num_inference_steps=20,
    generator=generator,
).images[0]
```

**Note** this requires both code changes inside the inpainting pipelines as well as requiring a new call input called `mask_latents`

@yiyixuxu could you maybe give this a try?

**Comment by yiyixuxu** [comment]
@patrickvonplaten 

I'm not sure how to make example 1 work in a way that will save time and memory.

currently we are doing `maske_image_latent = vae.encode(image * mask)` - I think it will NOT be the same as `vae.encode(image) * vae.encode(mask)`? so if we get the already computed latents for image and mask we will need to decode them? or is there a math that can make it work without decoding it?

example 2 should be fine, I will work on that first! 

YiYi

**Comment by patrickvonplaten** [comment]
Nice catch @yiyixuxu - I think you're right! Solution 1) might not work after all :-) Let's just focus on example 2 then!

**Comment by engrmusawarali** [comment]
@yiyixuxu @patrickvonplaten  @sayakpaul  How can we apply the mask in the latent space. Basically, I am creating a chain of inpainting pipelines. I dont want to decode the image.
